README
=encoding UTF-8

Sipwise-Base version 1.005

=head1 NAME

README - basic information for users prior to downloading

=head1 INSTALLATION

See L<http://www.cpan.org/modules/INSTALL.html>.

=head1 DEPENDENCIES

See distribution meta file.

=head1 LICENCE

This software is Copyright © 2013 by Sipwise GmbH, Austria.

This program is free software; you can redistribute it
and/or modify it under the terms of the GNU General Public
License as published by the Free Software Foundation; either
version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be
useful, but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU General Public License for more
details.

You should have received a copy of the GNU General Public
License along with this package; if not, write to the Free
Software Foundation, Inc., 51 Franklin St, Fifth Floor,
Boston, MA  02110-1301 USA

On Debian systems, the full text of the GNU General Public
License version 3 can be found in the file
F</usr/share/common-licenses/GPL-3>.


Changes
=encoding UTF-8

=head1 NAME

Changes - Revision history for Sipwise-Base

=head2 1.005  2013-09-12

=over

=item document release procedure

=item fix warning: Subroutine Sipwise::Base::CLASS redefined

=back

=head2 1.004  2013-07-04

=over

=item declare L<strictures> dependencies, only used for developers

=item Perl minimum version 5.14, brings changes to versioning and a Unicode handling aspect

=back

=head2 1.003  2013-06-17

=over

=item POD errors

=item spelling

=back

=head2 1.002  2013-04-29

=over

=item change licence to GPL3+

=item remove L<lambda> feature for easier packaging

=back

=head2 1.001  2013-04-23

=over

=item add support for selectively skipping certain features through import

=item coding standards advice about versions

=back

=head2 1.000  2013-03-21

Initial release.


t/Types.pm
package Types;
use Sipwise::Base;
use Moose::Util::TypeConstraints qw(subtype as where);

subtype 'Place', as 'Str', where { /world/ };


t/StrictSubs.pm
package StrictSubs;
use Sipwise::Base;
fnord;


t/StrictVars.pm
package StrictVars;
use Sipwise::Base;
$fnord;


Build.PL
use Module::Build qw();
my $builder = Module::Build->new(
    module_name       => 'Sipwise-Base',
    license           => 'GPL_3',
    dist_author       => 'Lars Dieckow <ldieckow@sipwise.com>',
    dist_version_from => 'lib/Sipwise/Base.pm',
    configure_requires => {
        'Module::Build' => '0.4004',
    },
    test_requires    => {
        'Moose::Util::TypeConstraints' => 0,
        'Test::DescribeMe' => 0,
        'Test::More' => '0.88',
    },
    requires => {
        'autodie'                    => 0,
        'indirect'                   => 'v0.30',
        'Module::Runtime'            => 0,
        'Moose'                      => 2,
        'MooseX::Method::Signatures' => 0,
        'parent'                     => 0,
        'perl'                       => '5.014',
        'perl5i::2'                  => 'v2.12',
        'strictures'                 => 1,
        'TryCatch'                   => 0,
        'utf8'                       => 0,
    },
    # these are actually <http://p3rl.org/CPAN::Meta::Spec#develop>,
    # but nothing consumes this yet
    recommends => {
        'bareword::filehandles' => 0,
        'indirect' => 0,
        'multidimensional' => 0,
        'Test::Spelling' => 0,
    },
    add_to_cleanup => ['Sipwise-Base-*'],
);
$builder->create_build_script;


lib/Sipwise/Boilerplate.pm
package Sipwise::Boilerplate 1.005;
use Sipwise::Base;

__END__

=encoding UTF-8

=head1 NAME

Sipwise::Boilerplate - a template for new code

=head1 SYNOPSIS

Copy the following text, outdent it, save to a new file, then fill it in. Delete the C<=for author> POD sections after
you are done with them.

    package Something::Or::Other 1.000;
    use Sipwise::Base;
    use namespace::sweep;

    … # attributes, methods, package-level code go here

    $CLASS->meta->make_immutable;

    __END__

    =encoding UTF-8

    =head1 NAME

    Something::Or::Other - [One line description of module's purpose here]

    =head1 VERSION

    This document describes Something::Or::Other version 1.000

    =head1 SYNOPSIS

        use Something::Or::Other qw();

        =for author to fill in:
            Brief code example(s) here showing commonest usage(s).
            This section will be as far as many users bother reading
            so make it as educational and exeplary as possible.

    =head1 DESCRIPTION

    =for author to fill in:
        Write a full description of the module and its features here.
        Use subsections (=head2, =head3) as appropriate.

    =head1 INTERFACE

    =for author to fill in:
        Write a separate section listing the public components of the modules
        interface. These normally consist of either subroutines that may be
        exported, or methods that may be called on objects belonging to the
        classes provided by the module.

    =head2 Composition

        Something::Or::Other
            ISA Something
            DOES OtherRole

    All methods and attributes not mentioned here are
    inherited from L<Something> or mixed in from L<OtherRole>.

    =head2 Attributes

    =head3 C<foo>

    Type C<Str>, B<required> attribute, designates the fnarg fnord.

    =head3 C<feh>

    Type C<Bool>, whether foo is encoded into fnord.

    =head2 Methods

    =head3 C<bar>

    Returns the bar fnord as string.

    =head3 C<mogrify>

        $soo->mogrify;  # use defaults
        $soo->mogrify({blah => 42, harbl => 23});

    Uses the supplied arguments to blah blah.

    =head4 C<mogrify> argument keys

    …

    =head2 Exports

    =head3 C<quux>

    Exported on demand, same as calling the default constructor followed by the L</bar> method.

    =head1 DIAGNOSTICS

    None.

    =for author to fill in:
        List every single error and warning message that the module can
        generate (even the ones that will "never happen"), with a full
        explanation of each problem, one or more likely causes, and any
        suggested remedies.

    =over

    =item C<< Error message here, perhaps with %s placeholders >>

    [Description of error here]

    =item C<< Another error message here >>

    [Description of error here]

    [Et cetera, et cetera]

    =back

    =head1 CONFIGURATION AND ENVIRONMENT

    Something::Or::Other requires no configuration files or environment variables.

    =for author to fill in:
        A full explanation of any configuration system(s) used by the
        module, including the names and locations of any configuration
        files, and the meaning of any environment variables or properties
        that can be set. These descriptions must also include details of any
        configuration language used.

    =head1 DEPENDENCIES

    See meta file in the source distribution.

    =for author to fill in:
        A list of all the other modules that this module relies upon,
        including any restrictions on versions, and an indication whether
        the module is part of the standard Perl distribution, part of the
        module's distribution, or must be installed separately.

    =head1 INCOMPATIBILITIES

    None reported.

    =for author to fill in:
        A list of any modules that this module cannot be used in conjunction
        with. This may be due to name conflicts in the interface, or
        competition for system or program resources, or due to internal
        limitations of Perl (for example, many modules that use source code
        filters are mutually incompatible).

    =head1 BUGS AND LIMITATIONS

    L<https://bugtracker.sipwise.com>

    No known limitations.

    =for author to fill in:
        A list of known problems with the module, together with some
        indication whether they are likely to be fixed in an upcoming
        release. Also a list of restrictions on the features the module
        does provide: data types that cannot be handled, performance issues
        and the circumstances in which they may arise, practical
        limitations on the size of data sets, special cases that are not
        (yet) handled, etc.

    =head1 TO DO

    =for author to fill in:
        future plans and ideas

    =head1 SEE ALSO

    =for author to fill in:
        related work and modules, perhaps with some prose describing the
        different advantages of each under which circumstances

    =head1 AUTHOR

    Your Name C<< <yname@sipwise.com> >>

    =head1 LICENCE

    restricted


lib/Sipwise/CodingStandards.pm
package Sipwise::CodingStandards 1.005;
use Sipwise::Base;

__END__

=encoding UTF-8

=head1 NAME

Sipwise::CodingStandards - Sipwise Perl coding standards

=head1 VERSION

This document describes Sipwise::CodingStandards version 1.005

=head1 DESCRIPTION

This document outlines the coding standards for new Perl projects. Their purpose is to guide programmers to a common
code style and practices, making decisions so you don't have to. Follow them unless you have a good reason not to. See
L<< chapter 1 in I<PBP>|http://oreilly.com/catalog/perlbp/ >> for the rationale. Taking Perl's philosophy to heart, the
standards are not enforced – it's up to the programmer to use common sense and sound judgement in applying them.

The standards are not set in stone, either. If you notice that consistently I<not> following them produces better
results, please communicate the matter and get this document ameliorated.

=head2 common modules

L<Sipwise::Base> is a module that simply bundles common other modules that you are very likely to include in a class
file - one line is enough to load all of them. The goal is to enable modern Perl programming and restrict some legacy
features of the language that should not be used anymore in new code.

=head2 versions

=head3 examples

For modules:

    package Foo::Bar 2.001; # 2nd version of revision 2
    package Fnord 2.420;    # 421st version of revision 2

Otherwise:

    our $VERSION = '2.001'; # 2nd version of revision 2
    our $VERSION = '2.420'; # 421st version of revision 2

=head3 rationale

Every module must have a version or else it is problematic to accurately specify a dependency when a distribution is
split. It works best when the distribution version and each module's version are equal and increased in lock-step. To
to make this easy, use C<perl-reversion>, which is part of L<Perl::Version>. Increase a version each time when you need
to depend on a new features or a changed API from code external to the distribution, it should be followed by a release
of the distribution.

When using the L<perlfunc/package> declaration, a version is a rational number. If a C<package> declaration is not
suitable, declare the magic C<$VERSION> variable (see L<perlobj/VERSION($need)>) and then a version is a string.

In any case, the version consists of a revision number, a literal dot as separator and a version number, altogether in
the form C<y.xxx>. C<y> is a natural number, C<xxx> are exactly three zero-padded digits. String quoting prevents
trailing zeroes from disappearing. Padding to the same amount of digits prevents confusion about C<< 1.10 < 1.9 >>.
Having exactly one separator prevents confusion about C<5.10.1 == 5.010001>.

Do not use the L<version> module for declaring versions. Using plain strings prevents that ugly C<v> prefix in a
distribution tarball name. However, using the module to handle versions, e.g. for comparing them, is a good idea.

Do not use L<v-strings|http://p3rl.org/data#Version-Strings>. They are poorly understood.

Do not use underscores. Using underscores require to L<perlfunc/eval> the version to turn it into a number. If you want
to mark a distribution as release candidate to the PAUSE indexer, add the word C<TRIAL> to the distribution name, see
L<http://pause.perl.org/pause/query?ACTION=pause_04about#developerreleases>.

=head3 compliance

This scheme is fully compatible with the advice laid out in L<perlmodstyle>, L<version>,
L<"strict" version rules|http://p3rl.org/perl5120delta#Version-number-formats>,
L<"Version numbers should be boring"|http://www.dagolden.com/index.php/369/>,
L<Perl::Critic::Policy::ValuesAndExpressions::ProhibitVersionStrings>,
L<Perl::Critic::Policy::ValuesAndExpressions::RequireNumericVersion>.

It is incompatible with L<< I<semver>|http://semver.org >>.

=head2 sample code template

See L<Sipwise::Boilerplate> in this distribution.

This template is derived from L<Module::Starter::PBP> for the most part.

=head2 C<perltidy>

See F<share/.perltidyrc> in this distribution.

L<perltidy> sometimes produces not optimal results with immediately nested pairs of types of brackets.

    $resultset->find({
            foo => 123,
        },
        {
            bar_id  => 23,
            quux_id => {-ident => 'baz_id'},
        },
    );

It is OK to manually fix this for symmetry.

    $resultset->find(
        {
            foo => 123,
        },
        {
            bar_id  => 23,
            quux_id => {-ident => 'baz_id'},
        },
    );

=head2 C<perlcritic>

See F<share/.perlcritic> in this distribution.

Install L<Task::Perl::Critic::IncludingOptionalDependencies> and L<Perl::Critic::logicLAB>.

=head1 AUTHOR

C<< <ldieckow@sipwise.com> >>


lib/Sipwise/Base.pm
package Sipwise::Base 1.005;
use 5.014;
use utf8;
use Module::Runtime qw(use_module);
use parent 'autodie';

sub import {
    my ($class, %param) = @_;
    my %features = (
        perl5i => q(use perl5i::2 -skip => [qw(Signatures Try::Tiny autodie)];),
        Moose => q(use Moose qw(after augment before extends has inner override super with);),
        'MooseX::Method::Signatures' => q(use MooseX::Method::Signatures;),
        TryCatch => q(use TryCatch;),
    );

    if ($param{-skip}) {
        for my $feature (@{ $param{-skip} }) {
            delete $features{$feature};
        }
    }
    my $caller = caller;

    use_module('strictures')->import(1);
    my @skip = qw(Signatures Try::Tiny autodie);
    push @skip, 'CLASS' if __PACKAGE__->can('CLASS');
    use_module('perl5i::2')->import(-skip => [@skip]);

    my $export = "package $caller;\n";
    for my $f (sort keys %features) {
        $export .= "$features{$f}\n";
    }
    eval $export;

    @_ = ($class, ':all');
    goto &autodie::import;
}

1;

__END__

=encoding UTF-8

=head1 NAME

Sipwise::Base - import boilerplate code

=head1 VERSION

This document describes Sipwise::Base version 1.005

=head1 SYNOPSIS

    use Sipwise::Base;

=head1 DESCRIPTION

The purpose of this module is to reduce the amount of boilerplate code and to enable some Perl features that are not
available by default. This is in the spirit of
L<http://modernperlbooks.com/mt/2009/10/remove-the-little-pessimizations.html|eliminating pessimisations>.

Switches on the functionality of the pragmatic modules L<strictures>, L<perl5i>, L<Moose>,
L<MooseX::Method::Signatures>, L<TryCatch>.

=head1 INTERFACE

=head2 C<import>

See L<perlfunc/import>.

By passing an arrayref to C<-skip>, you can disable features:

    use Sipwise::Base -skip => ['Moose'];   # skip importing Moose

The features strings are C<perl5i>, C<Moose>, C<MooseX::Method::Signatures>, C<TryCatch>.

=head2 Exports

=head1 DIAGNOSTICS

None.

=head1 CONFIGURATION AND ENVIRONMENT

Sipwise::Base requires no configuration files or environment variables.

=head1 DEPENDENCIES

See meta file in the source distribution.

=head1 INCOMPATIBILITIES

None reported.

=head1 BUGS AND LIMITATIONS

L<https://bugtracker.sipwise.com>

=head1 TO DO

different profiles for command-line applications

=head1 SEE ALSO

L<ToolSet>, L<Toolkit>, Ingy's L<perl5>, L<Syntax::Collector>

=head1 AUTHOR

Lars Dieckow C<< <ldieckow@sipwise.com> >>

=head1 LICENCE

This software is Copyright © 2013 by Sipwise GmbH, Austria.

This program is free software; you can redistribute it
and/or modify it under the terms of the GNU General Public
License as published by the Free Software Foundation; either
version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be
useful, but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU General Public License for more
details.

You should have received a copy of the GNU General Public
License along with this package; if not, write to the Free
Software Foundation, Inc., 51 Franklin St, Fifth Floor,
Boston, MA  02110-1301 USA

On Debian systems, the full text of the GNU General Public
License version 3 can be found in the file
F</usr/share/common-licenses/GPL-3>.
